src/tuner/algo/allreduce/ring.h

```diff
@@ -0,0 +1,311 @@
+{
+	return get_ring_chunk_steps(proto)*get_buffsize(proto)*get_overhead_numerator(proto)/get_overhead_denominator(proto)/NCCL_OFI_TUNER_NCCL_STEPS;
+}
+
+/**
+ * @brief Get the size of the ring "window" for a given protocol, number of channels, and number of
+ *        ranks.
+ *
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The size of the ring window.
+ */
+static inline double get_ring_size_window(unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+    return num_channels * num_ranks * get_ring_chunk_window_in_bytes(proto);
+}
+
+/**
+ * @brief Get the number of "windows" required for a given data size, protocol, number of channels,
+ *        and number of ranks. This is a term we made up for our tuner, NCCL may use a different
+ *        name to describe this.
+ *
+ * @param nbytes The data size in bytes.
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of windows required.
+ */
+static inline int get_ring_num_windows(unsigned long long nbytes, unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+    const double size_window = 1.0*get_ring_size_window(num_channels, num_ranks, proto);
+    const double iters = ceil(nbytes / size_window);
+    int int_iters = iters;
+    return int_iters + 1;
+
+}
+
+/**
+ * @brief Get the number of bytes remaining at the beginning of the ith iteration.
+ *
+ * @param i The iteration index.
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The number of bytes processed in the given iteration.
+ */
+static inline unsigned long long processed_bytes(unsigned long long nbytes, unsigned long long i, unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+    return i * get_ring_size_window(num_channels, num_ranks, proto);
+}
+
+/**
+ * @brief Get the bytes to be processed in remaining iterations, including the ith iteration.
+ *
+ * @param i The iteration index.
+ * @param nbytes The data size in bytes.
+ * @param num_channels The number of channels.
+ * @param num_ranks The number of ranks.
+ * @param proto The protocol (NCCL_PROTO_SIMPLE, NCCL_PROTO_LL, or NCCL_PROTO_LL128).
+ * @return The remaining bytes to be processed in the given iteration.
+ */
+static inline unsigned long long remaining_bytes(unsigned long long i, unsigned long long nbytes, unsigned long long num_channels, unsigned long long num_ranks, int proto)
+{
+	const unsigned long long done = processed_bytes(nbytes, i, num_channels, num_ranks, proto);
+	if (done < nbytes) {
+		return nbytes - done;
+	}
+	return 0;
```

Comment:
shouldn't we non-zero remaining when `done > nbytes`?

Reply:
no, because all return values and arguments to the function are unsigned